src/net/sourceforge/kolmafia/moods/MoodManager.java

```diff
@@ -155,9 +155,10 @@ public static MoodTrigger addTrigger(final String type, final String name, final
     return trigger;
   }
 
-  /** Removes all from the current displayList. */
+  /** Removes all triggers from the current mood and displayList. */
   public static void removeTriggers(final Collection<MoodTrigger> triggers) {
-    for (MoodTrigger trigger : triggers) {
+    MoodTrigger[] local = triggers.toArray(new MoodTrigger[0]);
```

Comment:
Surely there is a way to do this that doesn't require you to copy the input list. For example, using an explicit Iterator.

Reply:
Yes.  Is this conversation about "better" or just "different"?

Previously one of the removeTrigger routines was passed an array of Objects.  It iterated over the array and so there were different structures - one controlling the iteration and one holding the data being removed.  

The delinting that turned some arrays into lists created this bug.  The decision to copy to an array and iterate over that was made because it restored (a close approximation to) the previous implementation, in the hope that restoring the previous implementation was less likely to was less likely to introduce a new bug.  I also figured performance wasn't an issue since 1000 is certainly an upper bound on the size of the trigger list.

So the array copy was a deliberate choice but if I'm the only one who thinks so then an iterator will be used.

Reply:
After a good nights sleep....

The "contract" for  removeTriggers is that it accepts a collection or triggers and then deletes any trigger that is in the collection from the (should be) internal list of mood triggers.  As such there is no requirement that the collection be a subset of the mood triggers so using the collection iterator to delete from the internal list is not going to work if the collection is just an arbitrary list of triggers.

So if we are going to allow removeTriggers to accept an arbitrary collection then we cannot use the iterator to delete.  So a defensive copy is appropriate.

I can imagine new KoLmafia code that makes a list of triggers that cannot be used and removing all those triggers.  It might allow for some automation when ascending.  But no one has written that yet.  

I think the core problem is that the mood command "knows" that there are functions to get a list of all triggers and to delete all the triggers in a list and composes them tom delete all triggers.  It seems like a "delete all triggers" function would expose not as many details of the MoodManager and that might be the best way to go.

Reply:
You are right - the method cannot "know" that the collection it is passed is the internal list of triggers. But it can detect when it is - since it is explicitly deleting from that list - and use the iterator to delete, in that case.

`  /** Removes all from the current displayList. */
  public static void removeTriggers(final Collection<MoodTrigger> triggers) {
    Iterator<MoodTrigger> it = triggers.iterator();
    while (it.hasNext()) {
      MoodTrigger trigger = it.next();
      if (MoodManager.currentMood.removeTrigger(trigger)) {
        if (triggers == MoodManager.displayList) {
          it.remove();
        } else {
          MoodManager.displayList.remove(trigger);
        }
      }
    }
  }
`

Reply:
Regarding "better" vs. "different":

Method A iterates through a list, search/removing elements from two other comparable lists. It copies the input list. So, the algorithm is O(n squared) in computation and O(n) in memory.

Method B does the same with no data copying. It is O(0) in memory.

They are "different". Is one "better"? Seemed obvious to me, which is why I phrased my suggestion in terms of "without allocating memory"